### PR TITLE
Field encryption curated sample

### PR DESCRIPTION
**Pull request by clementetb**
Adds the Curated examples project, a set of samples to show different patterns and uses cases around Realm, Sync, and Atlas. 

As part of this initial work we also introduced the first example, Field level encryption.

### CODE REVIEW

**Review by cmelchior — COMMENTED** [review]
Initial batch of comments. Mostly for the docs.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ext/EncryptionExt.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import java.security.Key
+import java.security.MessageDigest
+import javax.crypto.KeyGenerator
+
+const val HASH_ALGORITHM = "SHA-256"
+
+fun Key.computeHash(): ByteArray =
+    MessageDigest.getInstance(HASH_ALGORITHM).digest(encoded)
+
+suspend fun UserKeyStore.getKeyOrGenerate(
```
Comment (cmelchior) [inline CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ext/EncryptionExt.kt:32]: Why are these extension methods? 

Also, it would be nice with some documentation for these methods.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ext/UserExt.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import io.realm.curatedsyncexamples.fieldencryption.models.SerializableCipherSpec
+import io.realm.curatedsyncexamples.fieldencryption.models.CustomData
+import io.realm.curatedsyncexamples.fieldencryption.models.UserKeyStore
+import io.realm.kotlin.annotations.ExperimentalRealmSerializerApi
+import io.realm.kotlin.mongodb.User
+import io.realm.kotlin.mongodb.ext.call
+import io.realm.kotlin.mongodb.ext.customData
+
+@OptIn(ExperimentalRealmSerializerApi::class)
+fun User.keyStore(): UserKeyStore {
```
Comment (cmelchior) [inline CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ext/UserExt.kt:28]: These could benefit from some documentation.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/ui/DemoSelectorViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+class ExamplesScreenViewModel(private val apps: List<DemoWithApp>) : ViewModel() {
+    private val loading = MutableStateFlow(true)
+    val loadingState: StateFlow<Boolean> = loading.asStateFlow()
+
+    val demoEntriesWithStatus: MutableLiveData<List<DemoWithStatus>> by lazy { MutableLiveData<List<DemoWithStatus>>() }
+
+    private suspend fun App.isAvailable() =
+        try {
+            login(Credentials.anonymous())
```
Comment (cmelchior) [inline CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/ui/DemoSelectorViewModel.kt:44]: I think that this is maybe what is messing a bit with the logic when starting the app the first time. Shouldn't we log this user out again before returning?

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/ui/DemoSelectorViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            true
+        } catch (e: ServiceException) {
+            e.message?.startsWith("[Service][Unknown(4351)]") != true
```
Comment (cmelchior) [inline CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/ui/DemoSelectorViewModel.kt:47]: I feel like this would be a good fit for an `UnknownAppServiceException` or similar in the SDK? But this is fine for now.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ui/records/SecretRecordsViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    private lateinit var user: User
+
+    private val _uiState = MutableStateFlow(SecretRecordsUiState())
+    val uiState: StateFlow<SecretRecordsUiState> = _uiState.asStateFlow()
+
+    init {
+        viewModelScope.launch(Dispatchers.IO) {
+            user = app.currentUser!!
+            cipherSpec = user.fieldEncryptionCipherSpec()
+            runBlocking {
```
Comment (cmelchior) [inline CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ui/records/SecretRecordsViewModel.kt:63]: You shouldn't need a runBlocking inside a `launch` method?

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/models/SerializableCipherSpec.kt [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import java.lang.Exception
+import java.security.Key
+import javax.crypto.Cipher
+import javax.crypto.spec.IvParameterSpec
+
+private const val IV_SIZE: Int = 16  // should be ok for most of the cases
+
+/**
+ * Class that contains all the information required to instantiate a [Cipher], it also provides
+ * with methods to encrypt/decrypt data.
```
Comment (cmelchior) [inline CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/models/SerializableCipherSpec.kt:31]:
```suggestion
 * Class that contains all the information required to instantiate a [Cipher], it also provides
 * methods to encrypt/decrypt data.
```

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/models/SystemKeyStore.kt [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            .apply {
+                load(null)
+            }
+
+    fun containsKey(keyAlias: String) = keyStore.isKeyEntry(keyAlias)
+
+    fun removeKey(keyAlias: String) = keyStore.deleteEntry(keyAlias)
+
+    /**
+     * Suspend as it key computation can take some time.
```
Comment (cmelchior) [inline CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/models/SystemKeyStore.kt:44]:
```suggestion
     * Made suspendable as key computation can take some time.
```

**Review by cmelchior — COMMENTED** [review]
Really nice 🥇 

Mostly minor comments, but very minor.

My biggest concern is still that this is as fairly advanced implementation, so I think we really need to make the README very detailed with a step-by-step guide, pointing to the places in code where things are happening all the way from "login" -> "read/write data"

File: CuratedSyncExamples/apps/field-encryption/README.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
+# Field level encryption
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply CuratedSyncExamples/apps/field-encryption/README.md:1]: Agreed -- this is potentially confusing

File: CuratedSyncExamples/.gitignore [hunk 9]
```diff
@@ -0,0 +1,15 @@
+*.iml
+.gradle
+/local.properties
+/.idea/caches
```
[the remark this answers is not in the document]
Comment (clementetb) [reply CuratedSyncExamples/.gitignore:4]: It was the default gitignore config

File: CuratedSyncExamples/apps/field-encryption/README.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+The purpose of this demo is to show how can we leverage the MongoDB and  resources to provide end-to-end field level encryption to users with multi-device access. 
+
+## Overview
+
+The encryption would be done using the Android keystore system. It offers enhanced security in key handling, any cryptographic key stored in this container is protected from unathorized use. Once a key is in the Android Keystore it can be used for different cryptographic operations, but it is not exportable.
+
+If we only relied on the Android KeyStore to handle the encryption keys, the user would only had access to the data from a single device, keys won't be exportable into other devices. To overcome this issue we have introduced a user bound keystore in Atlas, it stores the user's keys, and it allows to import them into any new device.
+
+
+## User keystore
```
[the remark this answers is not in the document]
Comment (clementetb) [reply CuratedSyncExamples/apps/field-encryption/README.md]: Should be more comparable now.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/Constants.kt [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.curatedsyncexamples
+
+const val FIELD_ENCRYPTION_APP_ID = "field-encryption-fjrvt"
```
[the remark this answers is not in the document]
Comment (clementetb) [reply CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/Constants.kt:19]: No, I will remove it.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ext/EncryptionExt.kt (hunk 1 above)
Comment (clementetb) [reply CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ext/EncryptionExt.kt]: No longer applicable, we have removed the UserKeyStore.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ui/records/SecretRecordsViewModel.kt (hunk 5 above)
Comment (clementetb) [reply CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/fieldencryption/ui/records/SecretRecordsViewModel.kt]: No, it is not needed. I moved some code and forgot to remove it.

File: CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/ui/DemoSelectorViewModel.kt (hunk 3 above)
Comment (clementetb) [reply CuratedSyncExamples/demo/src/main/java/io/realm/curatedsyncexamples/ui/DemoSelectorViewModel.kt]: Yes, that is causing issues. I am going to use `resendConfirmationEmail`

File: CuratedSyncExamples/README.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# Kotlin SDK Curated samples
```
[the remark this answers is not in the document]
Comment (clementetb) [reply CuratedSyncExamples/README.md:1]: Updated to "App Services Usage Samples"

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/records/SecretRecordsViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    private val _uiState = MutableStateFlow(SecretRecordsUiState())
+    val uiState: StateFlow<SecretRecordsUiState> = _uiState.asStateFlow()
+
+    init {
+        viewModelScope.launch(Dispatchers.IO) {
+            app.currentUser?.let { user ->
+                this@SecretRecordsViewModel.user = user
+
+                // Import key and cipher spec required to access the secret contents
+                PLECipherSpec = user.getPropertyEncryptionCipherSpec()
```
Comment (rorbech) [inline AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/records/SecretRecordsViewModel.kt:70]: Maybe clear those once the user logs out

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/records/SecretRecordsScreen.kt [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+import kotlin.io.encoding.Base64
+import kotlin.io.encoding.ExperimentalEncodingApi
+
+/**
+ * This screen would display a list of [SecretRecord] and a control view to add new entries. For each
+ * entry it would show its encrypted content in hexadecimal and the uncrypted value.
+ */
+@OptIn(ExperimentalEncodingApi::class)
+@Composable
+fun SecretRecordScreen(
```
Comment (rorbech) [inline AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/records/SecretRecordsScreen.kt:65]: Keyboard pops up when you focus the input field for the record and this somehow causes the first entry to not show up until the keyboard is hidden again. A bit annoying since you don't think that you added anything.

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/login/LoginViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+    fun login(email: String, password: String, register: Boolean = false) {
+        _uiState.update {
+            it.copy(loggingIn = true, errorMessage = null)
+        }
+
+        viewModelScope.launch(Dispatchers.IO) {
+            try {
+                if (register) {
+                    app.emailPasswordAuth.registerUser(email, password)
```
Comment (rorbech) [inline AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/login/LoginViewModel.kt:56]: Nit pick, so just leave out if not feeling for it: Should we ask the user to confirm the password when registering. A minor thing as you could just create another user or wipe it on the server, but I managed to type something that I couldn't repeat on second login 🤪

Comment (rorbech) [reply AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/login/LoginViewModel.kt:56]: Similar for the keystore password upon key generation

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/models/CustomData.kt [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+import kotlinx.serialization.Serializable
+
+/**
+ * App services user custom data. It contains the resources required to achieve property level
+ * encryption:
+ * 1. Encryption key, stored and password-protected in [keyStore].
+ * 2. Field encryption cipher spec, defined in [PLECipherSpec],
+ */
+@Serializable
+class CustomData(
```
Comment (rorbech) [inline AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/models/CustomData.kt:29]: Should we change the name so it is explicit that this a specific interpretation of the user's custom data specific to this app. Maybe `CustomKeyConfiguration` ... #naming

File: AppServicesUsageSamples/apps/property-encryption/README.md [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+For this demo, we will use the Android Keystore available from SDK version > 18. It offers hardware-enhanced security like biometric authentication and guarantees that the cryptographic material stored in it is not exportable.
+
+We will use the symmetric encryption algorithm AES, as it provides great performance when encrypting files and data.
+
+## Multi-device support
+As we mentioned, the local device keystore does not support exporting keys, this means that a user would not be able to access their data from a new device as the encryption keys would only be available from the original one. To overcome this we need a shared keystore that would allow accessing these keys from any device. 
+
+![alt text](diagram1.svg "Deployment")
+
+For this sample, we store the user keys in a password-protected Keystore in the user's `CustomData`. This allows you to control access to these (encrypted) keys based on the permission system in App Services.
```
Comment (rorbech) [inline AppServicesUsageSamples/apps/property-encryption/README.md:28]: Should we reference the `initializeCustomData.js` and `updateKeyStore.js` functions here to give some direction of where this is handled?

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/Constants.kt [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.appservicesusagesamples
+
+const val PROPERTY_ENCRYPTION_APP_ID = "<insert-app-id>"
```
Comment (rorbech) [inline AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/Constants.kt:19]: Maybe a comment about what these constants refers to

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/login/LoginScreen.kt [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+import androidx.compose.ui.text.input.PasswordVisualTransformation
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import androidx.lifecycle.compose.collectAsStateWithLifecycle
+import org.koin.compose.koinInject
+
+/**
+ * This screens prompts the user to wheter register or authenticate via email-password.
```
Comment (rorbech) [inline AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/ui/login/LoginScreen.kt:61]:
```suggestion
 * This screens prompts the user to register or authenticate via email-password.
```

File: AppServicesUsageSamples/README.md [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+- Demo - Android app containing the different samples. Samples have been separated in different packages.
+- Apps - Atlas App services apps required by each sample.
+
+## Getting started
+
+The demos are indendepent of each other, this means that it is not required to install all the app services app samples to test an individual sample.
+
+To begin, locate the App services app sources that you wish to install. We have conveniently linked them in the Samples list of this document.
+
+Next, follow the steps outlined in the [Atlas documentation](https://www.mongodb.com/docs/atlas/app-services/apps/create/) to setup the apps. These docs will guide you through the process and help troubleshoot any issue you might encounter.
```
Comment (rorbech) [inline AppServicesUsageSamples/README.md:24]: I think this is a little sparse. At least I had troubles just getting up and running. Maybe we need to elaborate on the steps and show the actual import commands, or at at least point to some elaborate instructions, because simply creating an app and trying to upload the app didn't do the trick for me. Lets discuss offline.

**Review by rorbech — COMMENTED** [review]
Awesome infrastructure and very neat sample 🎉  I have some very minor comments and then we should discuss if we could improve the instructions of setting up/importing the Atlas App Services App.

File: AppServicesUsageSamples/apps/property-encryption/README.md [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+We have chosen `BKS` as the remote keystore format as it supports storing symmetric keys. It is also supported by the Android SDK and allows us to show the interaction between the remote and local keystores in the sample code.
+
+## Key import
+After the user logs in on a new device, we need to import any required keys from the remote keystore into the local. If no key were available in the remote keystore, we would generate a new one that will be stored both remotely and locally.
+
+![alt text](diagram2.svg "Key import")
```
[the remark this answers is not in the document]
Comment (clementetb) [reply AppServicesUsageSamples/apps/property-encryption/README.md:35]: Yes, I wanted to name the sequentially..

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/models/CustomData.kt (hunk 16 above)
Comment (clementetb) [reply AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/models/CustomData.kt:29]: `PropertyEncryptionConfiguration`

File: AppServicesUsageSamples/apps/property-encryption/README.md (hunk 17 above)
Comment (clementetb) [reply AppServicesUsageSamples/apps/property-encryption/README.md:28]: Great point!

File: AppServicesUsageSamples/README.md [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+- `demo` - Android app containing the different samples. Samples have been separated in different packages.
+- `apps` - Atlas App services apps required by each sample.
+
+## Getting started
+
+The demos are indendepent of each other, which means that it is not required to install all the app services app samples to test an individual sample.
+
+To get started, locate the App services app sources in `apps` that you wish to install. We have conveniently linked them in the Samples list of this document.
+
+Next, follow the steps outlined in the [Atlas documentation](https://www.mongodb.com/docs/atlas/app-services/apps/create/) to setup an Atlas account and `realm-cli`, skip the app creation steps, then create and import the app with the following commands:
```
Comment (rorbech) [inline AppServicesUsageSamples/README.md:24]: Better with the explicit import command, but still think it requires a bit of insights before the users can do it. 

I would rephrase it to very explicit steps. Something like: 
- Follow the https://www.mongodb.com/docs/atlas/app-services/apps/create/#before-you-begin steps
- Create a "Cluster0" 
- Login
- Import

If the "Cluster0" is not created the import will not work.

**Review by cmelchior — COMMENTED** [review]
This is shaping up to be really great 🥇. Only have some minor comments.

File: AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/models/SecretRecord.kt [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+class SecretRecord : RealmObject {
+    @PersistedName("owner_id")
+    var ownerId: String = ""
+    @PrimaryKey
+    var _id: BsonObjectId = BsonObjectId()
+
+    /**
+     * Contains encrypted data.
+     */
+    var securedContent: ByteArray = byteArrayOf()
```
[the remark this answers is not in the document]
Comment (clementetb) [reply AppServicesUsageSamples/demo/src/main/java/io/realm/appservicesusagesamples/propertyencryption/models/SecretRecord.kt:38]: It cannot be private because then it won't be accessible by the delegate.

File: AppServicesUsageSamples/README.md [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+realm-cli login --api-key <your new public key> --private-api-key <your new private key>
+
+# Move to the app directory
+cd app/[APP-NAME]
+
+# Import the current app directory
+realm-cli import 
+```
+
+The `realm-cli import` command would prompt for the app configuration details, the default parameters should be right for most cases. Then the tool would publish the app and return the app-id for the newly creted app. Example:
```
Comment (cmelchior) [inline AppServicesUsageSamples/README.md:39]:
```suggestion
The `realm-cli import` command will prompt for the app configuration details, the default parameters should be suitable for most cases. Then the tool will publish the app and return the app-id for the newly created app. Example:
```

**Review by cmelchior — APPROVED** [review]
